pkg/scheduler/objects/application.go

```diff
@@ -57,6 +58,7 @@ var (
 )
 var initAppLogOnce sync.Once
 var rateLimitedAppLog *log.RateLimitedLogger
+var appSnapshotLock locking.RWMutex
```

Comment:
I'd pull this mutex inside the application so each app has a guard on the buffer and that is it
This way, we are serializing snapshots, but we don't need to do that. We just need to guard the buffer while serializing for each app
Each mutex is very small so it would be much better for the performance

Reply:
Excellent suggestion and much better than my package-level locks - fixed in commit 00e7935.